snuba/schemas.py

```diff
@@ -217,42 +183,149 @@ def get_time_series_query_schema_properties(default_granularity: int, default_wi
+        },
+    },
     # Need to select down to the project level for customer isolation and performance
     'required': ['project'],
+    'additionalProperties': False,
 }
 
-EVENTS_QUERY_SCHEMA['properties'].update({
-    **get_time_series_query_schema_properties(
+
+def get_time_series_extension_properties(default_granularity: int, default_window: timedelta):
+    return {
+        'type': 'object',
+        'properties': {
+            'from_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: (datetime.utcnow().replace(microsecond=0) - default_window).isoformat()
+            },
+            'to_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: datetime.utcnow().replace(microsecond=0).isoformat()
+            },
+            'granularity': {
+                'type': 'number',
+                'default': default_granularity,
+            },
+        },
+        'additionalProperties': False,
+    }
+
+
+@dataclass
+class Request:
+    query: Mapping[str, Any]
+    extensions: Mapping[str, Mapping[str, Any]]
+
+    def __post_init__(self):
+        self.body = ChainMap(self.query, *self.extensions.values())
+
+
+class RequestSchema:
+    def __init__(self, query_schema, extensions_schemas: Mapping[str, Any]):
+        self.__query_schema = query_schema
+        self.__extension_schemas = extensions_schemas
+
+        self.__composite_schema = {
+            'type': 'object',
+            'properties': {},
+            'required': [],
+            'definitions': {},
+            'additionalProperties': False,
+        }
+
+        for schema in itertools.chain([self.__query_schema], self.__extension_schemas.values()):
+            assert schema['type'] == 'object', 'subschema must be object'
+            assert schema['additionalProperties'] is False, 'subschema must not allow additional properties'
+            self.__composite_schema['required'].extend(schema.get('required', []))
+
+            for property_name, property_schema in schema['properties'].items():
+                assert property_name not in self.__composite_schema['properties'], 'subschema cannot redefine property'
+                self.__composite_schema['properties'][property_name] = property_schema
+
+            for definition_name, definition_schema in schema.get('definitions', {}).items():
+                assert definition_name not in self.__composite_schema['definitions'], 'subschema cannot redefine definition'
+                self.__composite_schema['definitions'][definition_name] = definition_schema
+
+        self.__composite_schema['required'] = set(self.__composite_schema['required'])
+
+    def validate(self, value) -> Request:
+        # XXX: Mutates input value!
+        validate(value, self.__composite_schema)
+
+        query = {key: value.pop(key) for key in self.__query_schema['properties'].keys() if key in value}
```

Comment:
What about we supported a query schema like this?

```
{
    main_field1: value,
    main_field2: value,
    timeseries_extension: {
         time_series_value1: ....
         ...
     },
    performance_extension: {
        ....
       ....
    }
}
```

While this is not backward compatible, thus we would have to support both formats for a while, this format would have two advantages:
1) semantics: we would not even have to impose the fields defined by extensions are disjoint. That would be structurally enforced.
2) The json validation of the extensions would be more sound, since we could just take the extension object and validate it against its extension schema instead of building the composite schema on the fly.

The drawback is that we would commit ourselves to disjoint fields. Which I am not sure it is a bad idea, I cannot think about a strong case for allowing an extension to override a field defined by another one (in terms of validation) that would justify the added complexity of defining the override semantics.

Reply:
Yes, this is my longer term plan, though I would prefer to structure it as:

```
{
  "query": {…},  # uses generic schema
  "extensions": {
    "name": {…},  # uses an extension schema
    "name": {…}   # uses another extension schema
  }
}
```

(Note that this is intentionally reflective of the structure of the `Request` dataclass.)

> semantics: we would not even have to impose the fields defined by extensions are disjoint. That would be structurally enforced.

I agree with this.

> I think we'll still want the composite schema, however, as it makes the validation error easier to understand since it's more reflective of the input in totality instead of independently validating each chunk.

I think we'll still want the composite schema, however, as it makes the validation error easier to understand since it's more reflective of the input in totality instead of independently validating each chunk.

> The drawback is that we would commit ourselves to disjoint fields. Which I am not sure it is a bad idea, I cannot think about a strong case for allowing an extension to override a field defined by another one (in terms of validation) that would justify the added complexity of defining the override semantics.

I think disjoint fields are only an artifact of the current API/schema implementation, as soon as we move to a new API/schema implementation and remove the old one, we can remove that requirement since it no longer has any benefit.

In my opinion extensions should be considered totally independent — when they are used in query evaluation, they should only get the generic query data as well as their extension parameters, and not have any visibility in to parameters provided to other extensions. (It'll take a little while to get there but that's at least how I think we should move forward.)

Reply:
ok,